docs/architecture/adr-050-sign-mode-textual.md

```diff
@@ -23,72 +24,119 @@ Protobuf-based SIGN_MODE_DIRECT was introduced in [ADR-020](./adr-020-protobuf-t
 
-In an effort to remove Amino from the SDK, a new sign mode needs to be created for hardware devices. [Initial discussions](https://github.com/cosmos/cosmos-sdk/issues/6513) propose a string-based sign mode, which this ADR formally specifies.
+In an effort to remove Amino from the SDK, a new sign mode needs to be created for hardware devices. [Initial discussions](https://github.com/cosmos/cosmos-sdk/issues/6513) propose a text-based sign mode, which this ADR formally specifies.
 
 ## Decision
 
-We propose to have SIGN_MODE_TEXTUAL’s signing payload `SignDocTextual` to be an array of strings, encoded as a `\n`-delimited string (see point #9). Each string corresponds to one "screen" on the hardware wallet device, with no (or little) additional formatting done by the hardware wallet itself.
+In SIGN_MODE_TEXTUAL, a transaction is rendered into a textual representation,
+which is then sent to a secure device or subsystem for the user to review and sign.
+Unlike `SIGN_MODE_DIRECT`, the transmitted data can be simply decoded into legible text
+even on devices with limited processing and display.
 
-```protobuf
-message SignDocTextual {
-  repeated string screens = 1;
-}
-```
+The textual representation is a sequence of _screens_.
+Each screen is meant to be displayed in its entirety (if possible) even on a small device like a Ledger.
+A screen is roughly equivalent to a short line of text.
+Large screens can be displayed in several pieces,
+much as long lines of text are wrapped,
+so no hard guidance is given, though 40 characters is a good target.
+A screen is used to display a single key/value pair for scalar values
+(or composite values with a compact notation, such as `Coins`)
+or to introduce or conclude a larger grouping.
+
+The text can contain the full range of Unicode code points, including control characters and nul.
+The device is responsible for deciding how to display characters it cannot render natively.
+See [annex 2](./adr-050-sign-mode-textual-annex2.md) for guidance.
 
-The string array MUST follow the specifications below.
+Screens have a non-negative indentation level to signal composite or nested structures.
+Indentation level zero is the top level.
+Indentation is displayed via some device-specific mechanism.
+Message quotation notation is an appropriate model, such as
+leading `>` characters or vertical bars on more capable displays.
 
-### 1. Bijectivity with Protobuf transactions
+Some screens are marked as _expert_ screens,
+meant to be displayed only if the viewer chooses to opt in for the extra detail.
+Expert screens are meant for information that is rarely useful,
+or needs to be present only for signature integrity (see below).
 
-The encoding and decoding operations between a Protobuf transaction (whose definition can be found [here](https://github.com/cosmos/cosmos-sdk/blob/master/proto/cosmos/tx/v1beta1/tx.proto#L13)) and the string array MUST be bijective.
+### Invertible Rendering
```

Comment:
What motivates the change in nomenclature away from "bijective"?

Reply:
Note that this draft PR is based off of the not-yet-committed #13434, so you're commenting on changes that will go away once this PR is rebased. You comments might be more appropriate for #13434.

But to quickly answer, "bijective" is not exactly correct. That would require that every valid Tx round-trip back to an equal Tx (yes) and also that every possible string round-trip back to an equal string (no). Having just the first condition makes `Format()` an "injection" (i.e. one-to-one), but this was deemed to be too technical, hence the use of "invertible".

Reply:
Should I move my comments to #13434, or is there enough visibility here?

Reply:
yes that's preferable. Let's keep this PR as message value renderer only

Reply:
Hmm - #13434 (structured screens) is just awaiting final approval, so I don't want to derail that. I'm not seeing anything in these comments that would make me want to add any late changes. This discussion seems more appropriate for #12910 - but as you can see that ship sailed a while ago. @peterbourgon it would be most appropriate if you could open an issue for discussion and raise you concerns there. I'm going to change the target for this draft to we can see the actual content of this PR better.